Avoid GAP namespaces and streamline code solving $x^2+y^2=-1$

fingolfin opened the pull request:
This is shorter and by removing a bunch of intermediate helpers arguably easier to understand. It is also a little bit faster this way.

fingolfin committed:
Avoid GAP namespaces and streamline code

This is shorter and by removing a bunch of intermediate
helpers arguably easier to understand.

olexandr-konovalov commented:
Thanks @fingolfin. Would be nice to have some tests to cover the patch. @drallenherman @angeldelriomateos do you have any suggestions?

fingolfin commented:
I'll add a test

fingolfin commented:
Tests added and everything works fine and has coverage.

fingolfin commented:
To be clear: I added tests for the function I changed, i.e. the one solving $x^2+y^2=-1$ over a finite field of odd order.

What is still missing is better / more tests for `PrimitiveIdempotentsNilpotent` which ensure that function ends up calling the function I changed here. As it is, a large part of that function is not exercised by the test suite at all.

fingolfin commented:
My new code has 93% code coverage (only two "error" statements are missing) and satisfies that documented API for the code it replaces. Anything else needed?

angeldelriomateos commented:
Dear Allen and Olexandr,
Thank you for your work.
Sorry for not being active with this.
All the best!
Ángel

fingolfin commented:
@angeldelriomateos no "Allen" is involved in this pull request, I think ;-).

Anyway: the changes here are really basic number theory. I teach this to my students in our "Elementary Number Theory" class, and Tonneli-Shanks in Cryptography -- but the kicker is that this PR actually *removes* the Tonneli-Shanks code, because it turns out it is not needed, as we can either just very rapidly find a square root of -1 by random guessing; or else the characteristic is congruent to 3 mod 4 and then we can directly compute square roots by powering.

So the code is much shorter, IMHO easier to understand, and even a bit more efficient.

olexandr-konovalov reviewed (approved):
Happy to merge this!
